### PR TITLE
ArchiSteamFarm.ui: use buildNpmPackage

### PR DESCRIPTION
**Pull request by legendofmiracles**
###### Description of changes

related to https://github.com/NixOS/nixpkgs/pull/229910

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/ArchiSteamFarm/web-ui/default.nix [hunk 1]
```diff
@@ -1,10 +1,8 @@
...
+{ lib, pkgs, fetchFromGitHub, buildNpmPackage, nodePackages, ArchiSteamFarm }:
 
-let
-  nodePackages = import ./node-composition.nix {
-    inherit pkgs nodejs;
-    inherit (stdenv.hostPlatform) system;
-  };
+buildNpmPackage {
+  pname = "asf-ui";
+  version = ArchiSteamFarm.version;
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/ArchiSteamFarm/web-ui/default.nix:5]:
```suggestion
  inherit (ArchiSteamFarm) version;
```

File: pkgs/applications/misc/ArchiSteamFarm/web-ui/default.nix [hunk 2]
```diff
@@ -1,10 +1,8 @@
-{ lib, pkgs, fetchFromGitHub, nodejs, nodePackages, stdenv, ArchiSteamFarm }:
+{ lib, pkgs, fetchFromGitHub, buildNpmPackage, nodePackages, ArchiSteamFarm }:
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/ArchiSteamFarm/web-ui/default.nix:1]: If we are no longer using node2nix, we should also drop pkgs from here

**Comment by legendofmiracles** [comment]
I'm testing it right now, and it isn't quite working yet. Give me a minute.

File: pkgs/applications/misc/ArchiSteamFarm/web-ui/default.nix [hunk 3]
```diff
@@ -15,20 +13,13 @@ let
...
+  nativeBuildInputs = [ nodePackages.node-gyp-build nodePackages.webpack-cli ];
 
-  postInstall = ''
-    patchShebangs node_modules/
-    npm run build
-    cp -r $out/lib/node_modules/asf-ui/dist $out/lib/dist
-    rm -rf $out/lib/node_modules/
+  installPhase = ''
+    mkdir $out
+    cp -rv dist/* $out/
```
Comment (mweinelt) [inline pkgs/applications/misc/ArchiSteamFarm/web-ui/default.nix:22]:
```suggestion
    runHook preInstall
    mkdir $out
    cp -rv dist/* $out/
    runHook preInstall
```

Always run pre and post hooks when overriding the phase.

File: pkgs/applications/misc/ArchiSteamFarm/web-ui/default.nix [hunk 4]
```diff
@@ -15,20 +13,13 @@ let
...
-in
-nodePackages.package.override {
-  inherit src;
-
-  # upstream isn't tagged, but we are using the latest official commit for that specific asf version (assuming both get updated at the same time)
-  version = ArchiSteamFarm.version;
+  npmDepsHash = "sha256-AY1DFuZkB8tOQd2FzHuNZ31rtLlWujP+3AqsMMB2BhU=";
 
-  nativeBuildInputs = [ pkgs.nodePackages.node-gyp-build ];
+  nativeBuildInputs = [ nodePackages.node-gyp-build nodePackages.webpack-cli ];
```
Comment (mweinelt) [inline pkgs/applications/misc/ArchiSteamFarm/web-ui/default.nix:18]: I'm surprised passing these packages is required at all.

Comment (legendofmiracles) [reply pkgs/applications/misc/ArchiSteamFarm/web-ui/default.nix]: You're right, fixed!

File: pkgs/applications/misc/ArchiSteamFarm/web-ui/default.nix [hunk 5]
```diff
@@ -15,20 +13,15 @@ let
...
-  nativeBuildInputs = [ pkgs.nodePackages.node-gyp-build ];
+    mkdir $out
+    cp -rv dist/* $out/
 
-  postInstall = ''
-    patchShebangs node_modules/
-    npm run build
-    cp -r $out/lib/node_modules/asf-ui/dist $out/lib/dist
-    rm -rf $out/lib/node_modules/
+    runHook preInstall
```
Comment (mweinelt) [inline pkgs/applications/misc/ArchiSteamFarm/web-ui/default.nix:24]: postInstall here :)

Comment (legendofmiracles) [reply pkgs/applications/misc/ArchiSteamFarm/web-ui/default.nix]: Good eyes 😅

**Comment by SuperSandro2000** [comment]
Tested and the output is identical, so nothing changed about it's working.

**Comment by SuperSandro2000** [comment]
we forgot to update the update script...

**Pull request merged**